hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/traceability/OpcodeTracer.java

```diff
@@ -0,0 +1,161 @@
+        this.context = ctx;
+        options = ctx.getOpcodeTracerOptions();
+        contractActions = ctx.getContractActions();
+        if (CollectionUtils.isEmpty(contractActions)) {
+            log.warn("No contract actions found in context!");
+        }
+    }
+
+    @Override
+    public void tracePostExecution(final MessageFrame frame, final Operation.OperationResult operationResult) {
+        final List<Bytes> memory = captureMemory(frame);
+        final List<Bytes> stack = captureStack(frame);
+        final Map<Bytes, Bytes> storage = captureStorage(frame);
+        opcodes.add(new Opcode(
+                frame.getPC(),
+                frame.getCurrentOperation().getName(),
+                frame.getRemainingGas(),
+                operationResult.getGasCost(),
+                frame.getDepth(),
+                stack,
+                memory,
+                storage,
+                frame.getRevertReason().map(Bytes::toString).orElse(null)));
+    }
+
+    @Override
+    public void tracePrecompileCall(final MessageFrame frame, final long gasRequirement, final Bytes output) {
+        final Optional<Bytes> revertReason =
+                frame.getRevertReason().isPresent() ? frame.getRevertReason() : getRevertReason(contractActions);
+
+        revertReason.ifPresent(bytes -> log.trace("Revert reason: {}", bytes.toHexString()));
+
+        opcodes.add(new Opcode(
+                frame.getPC(),
+                frame.getCurrentOperation() != null
+                        ? frame.getCurrentOperation().getName()
+                        : StringUtils.EMPTY,
+                frame.getRemainingGas(),
+                output != null ? gasRequirement : 0L,
+                frame.getDepth(),
+                Collections.emptyList(),
+                Collections.emptyList(),
+                Collections.emptyMap(),
+                revertReason.map(Bytes::toString).orElse(null)));
+    }
+
+    @Override
+    public void finalizeOperation(final MessageFrame frame) {
+        context.setOpcodes(opcodes);
+    }
+
+    private List<Bytes> captureMemory(final MessageFrame frame) {
+        if (!options.isMemory()) {
+            return Collections.emptyList();
+        }
+
+        final Bytes[] memoryContents = new Bytes[frame.memoryWordSize()];
+        for (int i = 0; i < memoryContents.length; i++) {
+            memoryContents[i] = frame.readMemory(i * 32L, 32);
+        }
+        return Arrays.asList(memoryContents);
+    }
+
+    private List<Bytes> captureStack(final MessageFrame frame) {
+        if (!options.isStack()) {
+            return Collections.emptyList();
+        }
+
+        final Bytes[] stackContents = new Bytes[frame.stackSize()];
+        for (int i = 0; i < stackContents.length; i++) {
+            // Record stack contents in reverse
+            stackContents[i] = frame.getStackItem(stackContents.length - i - 1);
+        }
+        return Arrays.asList(stackContents);
```

Comment:
Why are we creating an array just to convert it to a list?

```suggestion
        int size = frame.stackSize();
        var stack = new ArrayList<Bytes>(size);

        for (int i = 0; i < size; ++i) {
            stack.add(frame.getStackItem(size - 1 - i));
        }

        return stack;
```